Creating local test for find_hung_process.py

Elley-Folks opened the pull request:
## Reason
The additions will help the find_hung_process.py script be completely tested, even with the logic that would normally only be used on the job queuing system. The goal will be for the test to actually run the script, and check if the output matches what is expected.

## Design
This refs #24929, added the files so the find_hung_process.py script can be tested by run_tests. Added a temporary "tester" directory where some of the necessary files have been put for now. This includes the mock qstat program (used to replace the real qstat command from the PBS job queuing system), the bad_mpi binary (this is the file that will hang, the script will be ran on this file for the test).

## Impact
The find hung process script can have all of its logic tested, and it will be isolated from any queuing systems.

Elley-Folks committed:
Added verbose and local test for testing script without PBS queuing system

--verbose or -v outputs some of the CL commands used in find_hung_process.py

--test-local parameter overrides the node name pattern to look for localhost instead of the naming convention for nodes on the HPC.

Elley-Folks committed:
Added in test sub parser, cleaned up comments.

The sub parser has the set of commands used for 'local' testing of the find_hung_process.py script without a job queuing system present.

Next: will add an argument and logic for running mock qstat command without adding it to path.

Elley-Folks committed:
WIP finished local test for find_hung_process.py, added files to testers directory

For find_hung_process.py added an alternate qstat command that runs the mock qstat file without adding it to path. It simply swaps out the command that the program runs if the -qp arg is passed in.

Refactored some arg parse commands and added a sub parser for the local testing. To use the "testing mode" of the program, type in "test" in the command line as one of the options.

Added new files FindHungPRocessTester.py and qstat. These will be used for setting up the actual test for the program.

The goal is to eventually be able to run the program locally with the tester and verify that the output is as expected.

Elley-Folks committed:
WIP Minimum tester set up for find_hung_process.py

They very very bare minimum for the find_hung_process.py tester.

Needed to add a static method, and a "tests" input file for run_tests to work.

Needed to add a simple_diffusion.i file as well for the tester to run the test.

Also added a temp directory called "tester" for the "tests" file. It is in test/tests/tester/tests

Elley-Folks committed:
WIP working on setting up test for test harness

Added qstat into tester directory.
Added badmpi into tester directory.
Modified FindHungProcessTester.py to use getCommand directly instead of runCommand. This works for now.

Small changes to find_hung_process,py.

Next will want to clone my repo onto rod, and then make the test work on a linux system. For it to run on mac there are some additional things that need to be done like fixing grep grepping extra lines, and setting up a stack tracer of some sort.

Elley-Folks committed:
Updated FindHungProcessTester.py to work on linux

Test currently does run on linux. Need to now add pattern recognition of output to verify that find_hung_process.py is working correctly.

As a note, sleep for 5 seconds to make sure that program is hung in a state that is expected. In the first few seconds bad_mpi.cpp does not have the expected output because of timers. There could be improvements later to make the program hang quicker, and stay in the hung state that we're looking for.

It may not be ideal to use pkill, but it does work for now since the name of the binary shouldn't change in the tester.

Elley-Folks committed:
WIP added pattern recognition to tests

Added expect_out to check the format of the output of the script for the test.

Added that this test currently is only for the linux platform. It currently only works on linux, it is a goal in the future for this to work on more platforms.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1646948/) on fa84edc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24932/site)

This comment will be updated on new commits.

Elley-Folks committed:
WIP Refactored string passed into getCommand.

Built the string up instead of grossly passing in one huge chunk.

May need to double check logic. Not sure if I need to check args passed in from command line and / or args in the command string.

Elley-Folks committed:
Fixed logic and added execute permission for binary

For now the tester will build up strings from an empty string. With how this is used it is okay for now. This can't take in input from the command line currently. Could add this in the future.

PR kept failing civet tests because it could not run the binary. Used chmod to fix this temporarily until better methods are used.

Elley-Folks committed:
Added bad_mpi source

This is the program that will hang a number <n> stacks deep.

The depth is specified with using the -s option. The "s" is short for stack. This program is used to test the accuracy of find_hung_process.py, and to also test the script when it is not on a job queuing system.

Later on this can be used to test the search algorithm in find_hung_process.py. The stack depth can be adjusted, and then the output from the script will tell you if it correctly narrowed in on the location of where the program actually hung (which stack).
